[pytorch][nnc] protocol classes to persist the context for compiled functions

ljk53 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #56852 [pytorch][nnc] mobile nnc backend skeleton
* **#56851 [pytorch][nnc] protocol classes to persist the context for compiled functions**



This is part of the changes to enable NNC AOT compilation for mobile.
At the end of the ahead-of-time compilation the compiler produces two sets of artifacts:
1. "compiled assembly code" - kernel functions in assembly format optimized for target platforms;
2. "compiled model" - regular TorchScript model that contains serialized parameters (weights/bias/etc) and invokes kernel functions via "handles" (name/version id/input & output specs/etc of the kernel functions).

This PR introduces a set of classes to represent kernel functions (a.k.a "handles"), which can be serialized/deserialized into/from the "compiled model" as an IValue.
Also introduces APIs to register/look-up "compiled assembly code".

Differential Revision: [D27921866](https://our.internmc.facebook.com/intern/diff/D27921866/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27921866/)!

Differential Revision: [D27921866](https://our.internmc.facebook.com/intern/diff/D27921866)

ljk53 committed:
[pytorch][nnc] protocol classes to persist the context for compiled functions

This is part of the changes to enable NNC AOT compilation for mobile.
At the end of the ahead-of-time compilation the compiler produces two sets of artifacts:
1. "compiled assembly code" - kernel functions in assembly format optimized for target platforms;
2. "compiled model" - regular TorchScript model that contains serialized parameters (weights/bias/etc) and invokes kernel functions via "handles" (name/version id/input & output specs/etc of the kernel functions).

This PR introduces a set of classes to represent kernel functions (a.k.a "handles"), which can be serialized/deserialized into/from the "compiled model" as an IValue.
Also introduces APIs to register/look-up "compiled assembly code".

Differential Revision: [D27921866](https://our.internmc.facebook.com/intern/diff/D27921866/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27921866/)!

[ghstack-poisoned]

ljk53 committed:
Update on "[pytorch][nnc] protocol classes to persist the context for compiled functions"


This is part of the changes to enable NNC AOT compilation for mobile.
At the end of the ahead-of-time compilation the compiler produces two sets of artifacts:
1. "compiled assembly code" - kernel functions in assembly format optimized for target platforms;
2. "compiled model" - regular TorchScript model that contains serialized parameters (weights/bias/etc) and invokes kernel functions via "handles" (name/version id/input & output specs/etc of the kernel functions).

This PR introduces a set of classes to represent kernel functions (a.k.a "handles"), which can be serialized/deserialized into/from the "compiled model" as an IValue.
Also introduces APIs to register/look-up "compiled assembly code".

Differential Revision: [D27921866](https://our.internmc.facebook.com/intern/diff/D27921866/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27921866/)!

[ghstack-poisoned]

ljk53 committed:
Update on "[pytorch][nnc] protocol classes to persist the context for compiled functions"


This is part of the changes to enable NNC AOT compilation for mobile.
At the end of the ahead-of-time compilation the compiler produces two sets of artifacts:
1. "compiled assembly code" - kernel functions in assembly format optimized for target platforms;
2. "compiled model" - regular TorchScript model that contains serialized parameters (weights/bias/etc) and invokes kernel functions via "handles" (name/version id/input & output specs/etc of the kernel functions).

This PR introduces a set of classes to represent kernel functions (a.k.a "handles"), which can be serialized/deserialized into/from the "compiled model" as an IValue.
Also introduces APIs to register/look-up "compiled assembly code".

Differential Revision: [D27921866](https://our.internmc.facebook.com/intern/diff/D27921866/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27921866/)!

[ghstack-poisoned]

ljk53 committed:
Update on "[pytorch][nnc] protocol classes to persist the context for compiled functions"


This is part of the changes to enable NNC AOT compilation for mobile.
At the end of the ahead-of-time compilation the compiler produces two sets of artifacts:
1. "compiled assembly code" - kernel functions in assembly format optimized for target platforms;
2. "compiled model" - regular TorchScript model that contains serialized parameters (weights/bias/etc) and invokes kernel functions via "handles" (name/version id/input & output specs/etc of the kernel functions).

This PR introduces a set of classes to represent kernel functions (a.k.a "handles"), which can be serialized/deserialized into/from the "compiled model" as an IValue.
Also introduces APIs to register/look-up "compiled assembly code".

Differential Revision: [D27921866](https://our.internmc.facebook.com/intern/diff/D27921866/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27921866/)!

[ghstack-poisoned]

ljk53 committed:
Update on "[pytorch][nnc] protocol classes to persist the context for compiled functions"


This is part of the changes to enable NNC AOT compilation for mobile.
At the end of the ahead-of-time compilation the compiler produces two sets of artifacts:
1. "compiled assembly code" - kernel functions in assembly format optimized for target platforms;
2. "compiled model" - regular TorchScript model that contains serialized parameters (weights/bias/etc) and invokes kernel functions via "handles" (name/version id/input & output specs/etc of the kernel functions).

This PR introduces a set of classes to represent kernel functions (a.k.a "handles"), which can be serialized/deserialized into/from the "compiled model" as an IValue.
Also introduces APIs to register/look-up "compiled assembly code".

Differential Revision: [D27921866](https://our.internmc.facebook.com/intern/diff/D27921866/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27921866/)!

[ghstack-poisoned]

ljk53 committed:
Update on "[pytorch][nnc] protocol classes to persist the context for compiled functions"




This is part of the changes to enable NNC AOT compilation for mobile.
At the end of the ahead-of-time compilation the compiler produces two sets of artifacts:
1. "compiled assembly code" - kernel functions in assembly format optimized for target platforms;
2. "compiled model" - regular TorchScript model that contains serialized parameters (weights/bias/etc) and invokes kernel functions via "handles" (name/version id/input & output specs/etc of the kernel functions).

This PR introduces a set of classes to represent kernel functions (a.k.a "handles"), which can be serialized/deserialized into/from the "compiled model" as an IValue.
Also introduces APIs to register/look-up "compiled assembly code".

Differential Revision: [D27921866](https://our.internmc.facebook.com/intern/diff/D27921866/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27921866/)!

Differential Revision: [D27921866](https://our.internmc.facebook.com/intern/diff/D27921866)

[ghstack-poisoned]

ljk53 committed:
Update on "[pytorch][nnc] protocol classes to persist the context for compiled functions"




This is part of the changes to enable NNC AOT compilation for mobile.
At the end of the ahead-of-time compilation the compiler produces two sets of artifacts:
1. "compiled assembly code" - kernel functions in assembly format optimized for target platforms;
2. "compiled model" - regular TorchScript model that contains serialized parameters (weights/bias/etc) and invokes kernel functions via "handles" (name/version id/input & output specs/etc of the kernel functions).

This PR introduces a set of classes to represent kernel functions (a.k.a "handles"), which can be serialized/deserialized into/from the "compiled model" as an IValue.
Also introduces APIs to register/look-up "compiled assembly code".

Differential Revision: [D27921866](https://our.internmc.facebook.com/intern/diff/D27921866/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27921866/)!

Differential Revision: [D27921866](https://our.internmc.facebook.com/intern/diff/D27921866)

[ghstack-poisoned]

ljk53 committed:
Update on "[pytorch][nnc] protocol classes to persist the context for compiled functions"




This is part of the changes to enable NNC AOT compilation for mobile.
At the end of the ahead-of-time compilation the compiler produces two sets of artifacts:
1. "compiled assembly code" - kernel functions in assembly format optimized for target platforms;
2. "compiled model" - regular TorchScript model that contains serialized parameters (weights/bias/etc) and invokes kernel functions via "handles" (name/version id/input & output specs/etc of the kernel functions).

This PR introduces a set of classes to represent kernel functions (a.k.a "handles"), which can be serialized/deserialized into/from the "compiled model" as an IValue.
Also introduces APIs to register/look-up "compiled assembly code".

Differential Revision: [D27921866](https://our.internmc.facebook.com/intern/diff/D27921866/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27921866/)!

Differential Revision: [D27921866](https://our.internmc.facebook.com/intern/diff/D27921866)

[ghstack-poisoned]

ljk53 committed:
Update on "[pytorch][nnc] protocol classes to persist the context for compiled functions"




This is part of the changes to enable NNC AOT compilation for mobile.
At the end of the ahead-of-time compilation the compiler produces two sets of artifacts:
1. "compiled assembly code" - kernel functions in assembly format optimized for target platforms;
2. "compiled model" - regular TorchScript model that contains serialized parameters (weights/bias/etc) and invokes kernel functions via "handles" (name/version id/input & output specs/etc of the kernel functions).

This PR introduces a set of classes to represent kernel functions (a.k.a "handles"), which can be serialized/deserialized into/from the "compiled model" as an IValue.
Also introduces APIs to register/look-up "compiled assembly code".

Differential Revision: [D27921866](https://our.internmc.facebook.com/intern/diff/D27921866/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27921866/)!

Differential Revision: [D27921866](https://our.internmc.facebook.com/intern/diff/D27921866)

[ghstack-poisoned]

ljk53 committed:
Update on "[pytorch][nnc] protocol classes to persist the context for compiled functions"




This is part of the changes to enable NNC AOT compilation for mobile.
At the end of the ahead-of-time compilation the compiler produces two sets of artifacts:
1. "compiled assembly code" - kernel functions in assembly format optimized for target platforms;
2. "compiled model" - regular TorchScript model that contains serialized parameters (weights/bias/etc) and invokes kernel functions via "handles" (name/version id/input & output specs/etc of the kernel functions).

This PR introduces a set of classes to represent kernel functions (a.k.a "handles"), which can be serialized/deserialized into/from the "compiled model" as an IValue.
Also introduces APIs to register/look-up "compiled assembly code".

Differential Revision: [D27921866](https://our.internmc.facebook.com/intern/diff/D27921866/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27921866/)!

Differential Revision: [D27921866](https://our.internmc.facebook.com/intern/diff/D27921866)

[ghstack-poisoned]

ljk53 committed:
Update on "[pytorch][nnc] protocol classes to persist the context for compiled functions"




This is part of the changes to enable NNC AOT compilation for mobile.
At the end of the ahead-of-time compilation the compiler produces two sets of artifacts:
1. "compiled assembly code" - kernel functions in assembly format optimized for target platforms;
2. "compiled model" - regular TorchScript model that contains serialized parameters (weights/bias/etc) and invokes kernel functions via "handles" (name/version id/input & output specs/etc of the kernel functions).

This PR introduces a set of classes to represent kernel functions (a.k.a "handles"), which can be serialized/deserialized into/from the "compiled model" as an IValue.
Also introduces APIs to register/look-up "compiled assembly code".

Differential Revision: [D27921866](https://our.internmc.facebook.com/intern/diff/D27921866/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27921866/)!

Differential Revision: [D27921866](https://our.internmc.facebook.com/intern/diff/D27921866)

[ghstack-poisoned]

kimishpatel commented on torch/csrc/jit/mobile/nnc/context.h:
```diff
@@ -0,0 +1,206 @@
+  //   temporary buffer 1
+  //   ...
+  //   temporary buffer BUFFER_NUM
+  std::vector<void*> arguments_;
```
If arguments are void* can you add a comment to the affect. 
Also for my understanding. What is the temporary buffers? And why there are N of them?

kimishpatel replied:
Also, I assume this works right now without having to specify size because kernels are statically sized, but this is generic enough for us to get size information at the least? Meaning we can have pair of {tensor_data_ptr, tensor_size_ptr}?

ljk53 replied:
I explained temporary buffers in the comment above related the the MemoryPlan.

I agree with these comments - it's just to make the thing work end-to-end with the current AOT compiler prototype so people can iterate on this independently. I think we can make the interface more general in subsequent PRs. We are not even in alpha test so the interface can be changed without much burden.

kimishpatel commented on torch/csrc/jit/mobile/nnc/context.h:
```diff
@@ -0,0 +1,206 @@
+
+  void allocate(ExecutionState* state) const;
+
+  std::vector<int64_t> buffer_sizes_;
```
Do you also need type information? How do you know if you need to allocate byte array or float?

ljk53 replied:
Yes, these buffer sizes should be a function of dynamic shape information and dtype information - right now all these are fixed so these can be fixed here as well.
I don't know what's the ultimate form of the representation (need to discuss with @ZolotukhinM and @bertmaher), and that's why I want to store these in a separate MemoryPlan structure instead of on the Function directly - because I feel it will probably grow to something more complicated than just a vector of "ScratchBufferSizes".

kimishpatel commented on torch/csrc/jit/mobile/nnc/context.h:
```diff
@@ -0,0 +1,206 @@
+  mutable std::unique_ptr<ExecutionState> execution_state_;
+};
+
+// Represents a set of compiled NNC functions which has a 1-1 correspondence
```
Is the comment correct?

ljk53 replied:
I meant to say CompilationUnit consists of a set of compiled NNC functions, and CompilationUnit has a 1-1 correspondence with a Module.
Do you think this rephrase is correct? If so I'll update the comment to make it more clear.

kimishpatel replied:
I see. Yes that rephrasing makes is cleaner.

kimishpatel commented on torch/csrc/jit/mobile/nnc/context.cpp:
```diff
@@ -0,0 +1,252 @@
+
+  dict.insert("name", name_.qualifiedName());
+  dict.insert("nnc_kernel_id", nnc_kernel_id_);
+  dict.insert("parameters", parameters_);
```
Interesting that parameters are associated with function and not module. Why not have equivalent module that owns the parametser?

ljk53 replied:
Good question - I'll add a TODO here. I'll improve it when we have more than one methods to compile.
I think parameters should not only be persisted with module (so multiple functions can share) but also be shared between the compiled model and the original model.
I mentioned in the design doc that eventually the compiled model should be an addition to the original model (like lite-interpreter's bytecode is an addition to the original model). We just recently solved the parameter sharing problem for lite-interpreter, so it will take some effort.

kimishpatel replied:
> but also be shared between the compiled model and the original model.

Exactly right. I think this may require some change to backend interface for it to allow querying parameters in some "universal" space.

kimishpatel commented on torch/csrc/jit/mobile/nnc/context.cpp:
```diff
@@ -0,0 +1,252 @@
+  dict.insert("parameters", parameters_);
+
+  // input_specs_
+  std::vector<c10::IValue> inputs;
```
Can this be named `seraialized_input_specs` or something similar? `inputs` feels like they are actual inputs.

ljk53 replied:
Sure, I'll update.

kimishpatel commented on torch/csrc/jit/mobile/nnc/context.cpp:
```diff
@@ -0,0 +1,252 @@
+
+void MemoryPlan::allocate(ExecutionState* state) const {
+  auto& allocations = state->preallocations_;
+  allocations.reserve(buffer_sizes_.size());
```
Do you also need to clear? Not necessary here probably but if we call allocate multiple times.

ljk53 replied:
I'll add it.

kimishpatel commented on torch/csrc/jit/mobile/nnc/context.cpp:
```diff
@@ -0,0 +1,252 @@
+  for (size_t i = 0; i < inputs.size(); ++i) {
+    const c10::IValue& input = inputs[i];
+    const auto& input_tensor = input.toTensor();
+    TORCH_CHECK(
```
So all these checks can also be using DelegatedBackendException with appropriate debug handle? Not for now but when we integrate debug handles.

ljk53 replied:
Yes we can - but the model hasn't really started running, guess the most useful context is probably just the pointer to the compiled Module + Method?
This "run" function is called from the generated TorchScript wrapper code (executed by the standard lite-interpreter), where it should have the Module/Method sources as well. So it could also catch and re-throw with the debug-handle saved there.

kimishpatel replied:
Makes sense.

raziel replied:
Currently we assume the call succeeds: self.__backend.execute(self.__handles["$method"], typed_inputs)

Or else the backend triggers the DelegatedBackendException with the right source data.

@kimishpatel @iseeyuan  Can you please take a look to make sure that even if the backend fails with another exception we fail with some useful information like you describe above?
@kimishpatel  This should probably be part of your PR where you introduce the DelegatedBackendException

kimishpatel replied:
Maybe for backend we should use the other macro for delegate exceptions `TORCH_DELEGATED_BACKEND_THROW`? You can use the debug handle corresponding to the first node of the graph or just use sentinel value of -1. However the former would be better such that you can at least get some stack trace, if say you lower only some submodule to nnc. So generally speaking moving to DelegatedBackendException is what we should do even if the information is sparse so that everything flows through it. What do you think @raziel?

kimishpatel commented on test/mobile/nnc/test_context.cpp:
```diff
@@ -0,0 +1,156 @@
+
+MemoryPlan create_test_memory_plan(const std::vector<int64_t>& buffer_sizes) {
+  MemoryPlan memory_plan;
+  memory_plan.buffer_sizes_ = buffer_sizes;
```
Should memory plan have API where it can create buffer sizes off of list of tensors?

kimishpatel commented on test/mobile/nnc/test_context.cpp:
```diff
@@ -0,0 +1,156 @@
+  return memory_plan;
+}
+
+TEST(Function, ExecuteBadMul) {
```
Why is it call BadMul? I thought it will throw.

ljk53 replied:
Oh, "bad_mul" is the term inspired by a micro-benchmark that does multiply by n times of addition. It's "bad" as it's inefficient.
Let me rename it to something less confusing.

kimishpatel commented on test/mobile/nnc/test_context.cpp:
```diff
@@ -0,0 +1,156 @@
+  f.set_input_specs({create_test_input_spec({size})});
+  f.set_output_spec({create_test_output_spec({size})});
+  f.set_parameters({at::ones({1}, at::kInt).mul(n)});
+  f.set_memory_plan(create_test_memory_plan({sizeof(float) * size}));
```
So memory plan is really for temporary buffers, is it? If so should it be called like that? I keep getting confused by it.

ljk53 replied:
I agree the current MemoryPlan interface is very rough. It simply allocates buffers for all the intermediate tensors in the graph, regardless their liveness. As revealed by the initial benchmark, this approach has visibly higher peak memory watermark as expected.
I think we will need more sophisticated approach for sure. We need to discuss the details. Maybe we can use the cache allocator, or the MemoryPlanner of the static runtime, or just allocate a big chunk of continuous buffer and let the compiled assembly code decide how to use it.
So I introduced this relatively dummy MemoryPlan stub in this initial PR.

kimishpatel replied:
Sounds good. I tried profiling allocator https://github.com/pytorch/pytorch/blob/master/c10/mobile/CPUProfilingAllocator.h last half but did not get enough cycles to continue working on it. That one basically allocates a big arena and references off of it. For future reference.

kimishpatel commented on test/mobile/nnc/test_context.cpp:
```diff
@@ -0,0 +1,156 @@
+      at::ones({32, 1, 3, 3}, at::kFloat)
+  });
+  f.set_memory_plan(create_test_memory_plan({
+      // NOLINTNEXTLINE(cppcoreguidelines-avoid-magic-numbers)
```
Yeah, I feel it will be cleaner if we have `Function`  with API `getScratchSpaceSizes` or something like that, which can be used with create_test_memory_plan.

ljk53 replied:
So "ScratchSpaceSizes" is about how to allocate temp scratch spaces, right? In my mental model I refer to this as "memory plan" (that's why I save it into MemoryPlan, and attach MemoryPlan to Function) - MemoryPlan is essentially off-spin of "ScratchSpaceSizes" from "Function", lol.

kimishpatel commented on test/mobile/nnc/test_context.cpp:
```diff
@@ -0,0 +1,156 @@
+      at::ones({size * 2}, at::kFloat)
+  });
+  // NOLINTNEXTLINE(cppcoreguidelines-avoid-goto,hicpp-avoid-goto)
+  EXPECT_THROW(
```
Should it throw with specific message?

ljk53 replied:
It has specific message on the throwing side ("Invalid input at pos..."), I was too lazy to check it in the unit test.
I could add it.

kimishpatel replied:
Not too important for now.

kimishpatel reviewed (commented):
Looks pretty clean. I left some high level comments.

One question I have is about dynamic shape inference? In terms of allocating temporary buffer space, do you have any thoughts on how that can be done. It would require shape inference too but I presume it should not be as detailed because network structure is fixed. Only thing you need is for each intermediate buffer what is the equation for buffer size given the primary input size.

kimishpatel reviewed (approved):
Looks good to me. Some minor changes requested but rest looks great.

iseeyuan reviewed (approved):
LGTM. I left comments in the internal diff and they are addressed.

ljk53 committed:
Update on "[pytorch][nnc] protocol classes to persist the context for compiled functions"




This is part of the changes to enable NNC AOT compilation for mobile.
At the end of the ahead-of-time compilation the compiler produces two sets of artifacts:
1. "compiled assembly code" - kernel functions in assembly format optimized for target platforms;
2. "compiled model" - regular TorchScript model that contains serialized parameters (weights/bias/etc) and invokes kernel functions via "handles" (name/version id/input & output specs/etc of the kernel functions).

This PR introduces a set of classes to represent kernel functions (a.k.a "handles"), which can be serialized/deserialized into/from the "compiled model" as an IValue.
Also introduces APIs to register/look-up "compiled assembly code".

Differential Revision: [D27921866](https://our.internmc.facebook.com/intern/diff/D27921866/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27921866/)!

Differential Revision: [D27921866](https://our.internmc.facebook.com/intern/diff/D27921866)

[ghstack-poisoned]

raziel reviewed (approved):
Minor comments/questions.
Thanks!

ljk53 commented on torch/csrc/jit/mobile/nnc/context.h:
```diff
@@ -0,0 +1,207 @@
+    name_ = name;
+  }
+
+  // The unique id of the generated NNC kernel corresponding to the function.
```
> Just curious how "unique" is the ID (within or beyond the compilation unit)?

It should be globally unique, and ideally idempotent (compiling the same model with the same version of AoT compiler should produce the same ID). In the current AoT compiler prototype I used model name, model version, method name and hash of the generated llvm assembly - I'm not entirely sure how stable the llvm assembly is, so the idempotent part might not be fulfilled. The purpose of the hash is to have some strict compatibility check to make sure compiled model and compiled assembly code are 1-1 mapped. Maybe we can relax it to some manually maintained compiler version number instead.

ljk53 commented on torch/csrc/jit/mobile/nnc/context.h:
```diff
@@ -0,0 +1,207 @@
+  void register_function(std::unique_ptr<Function> fn);
+
+ private:
+  C10_NODISCARD Function* find_function(const c10::QualifiedName& qn) const;
```
> For my own education, why is C10_NODISCARD needed in all these places?

I added these to get rid of the internal fbcode lint warnings...

I just did a search in the codebase and found it's not being widely used. I read the comments in the original PR #20324 - looks like the intention was to workaround some compiler error rather than lint warning. So maybe we should make this a convention?

cc: @ezyang @swolchok @smessmer who might know better.

ezyang replied:
`[[nodiscard]]` is C++17 and we are C++14. Therefore use of it in our codebase is nonportable and we must use a macro. We can consider removing the macro when we move to C++17. 

If the question is whether or not we need the `[[nodiscard]]` in the first place, it is because many compilers have a lint that require you to mark something as nodiscard if there are no side effects. To shut up the lint, you have to add nodiscard in a lot of places. Honestly this is kind of goofy (you gotta do it EVERYWHERE) but the code is net better, so you might as well do it.

ljk53 commented on torch/csrc/jit/mobile/nnc/context.h:
```diff
@@ -0,0 +1,207 @@
+ private:
+  C10_NODISCARD Function* find_function(const c10::QualifiedName& qn) const;
+
+  std::vector<std::unique_ptr<Function>> functions_;
```
Yeah, I copied this from lite-interpreter. Could update it to map.

ljk53 commented on torch/csrc/jit/mobile/nnc/context.h:
```diff
@@ -0,0 +1,207 @@
+
+// Specify the requirements on an input tensor.
+// TODO: support input tensor with dynamic shape (PR #54982)
+struct TORCH_API InputSpec {
```
Yeah, maybe I should create a TensorSpec structure and reuse it here. I didn't want to make it too involved in the initial draft.

ljk53 committed:
Update on "[pytorch][nnc] protocol classes to persist the context for compiled functions"




This is part of the changes to enable NNC AOT compilation for mobile.
At the end of the ahead-of-time compilation the compiler produces two sets of artifacts:
1. "compiled assembly code" - kernel functions in assembly format optimized for target platforms;
2. "compiled model" - regular TorchScript model that contains serialized parameters (weights/bias/etc) and invokes kernel functions via "handles" (name/version id/input & output specs/etc of the kernel functions).

This PR introduces a set of classes to represent kernel functions (a.k.a "handles"), which can be serialized/deserialized into/from the "compiled model" as an IValue.
Also introduces APIs to register/look-up "compiled assembly code".

Differential Revision: [D27921866](https://our.internmc.facebook.com/intern/diff/D27921866/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27921866/)!

Differential Revision: [D27921866](https://our.internmc.facebook.com/intern/diff/D27921866)

[ghstack-poisoned]

ljk53 committed:
Update on "[pytorch][nnc] protocol classes to persist the context for compiled functions"




This is part of the changes to enable NNC AOT compilation for mobile.
At the end of the ahead-of-time compilation the compiler produces two sets of artifacts:
1. "compiled assembly code" - kernel functions in assembly format optimized for target platforms;
2. "compiled model" - regular TorchScript model that contains serialized parameters (weights/bias/etc) and invokes kernel functions via "handles" (name/version id/input & output specs/etc of the kernel functions).

This PR introduces a set of classes to represent kernel functions (a.k.a "handles"), which can be serialized/deserialized into/from the "compiled model" as an IValue.
Also introduces APIs to register/look-up "compiled assembly code".

Differential Revision: [D27921866](https://our.internmc.facebook.com/intern/diff/D27921866/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27921866/)!

Differential Revision: [D27921866](https://our.internmc.facebook.com/intern/diff/D27921866)

[ghstack-poisoned]

ljk53 committed:
Update on "[pytorch][nnc] protocol classes to persist the context for compiled functions"




This is part of the changes to enable NNC AOT compilation for mobile.
At the end of the ahead-of-time compilation the compiler produces two sets of artifacts:
1. "compiled assembly code" - kernel functions in assembly format optimized for target platforms;
2. "compiled model" - regular TorchScript model that contains serialized parameters (weights/bias/etc) and invokes kernel functions via "handles" (name/version id/input & output specs/etc of the kernel functions).

This PR introduces a set of classes to represent kernel functions (a.k.a "handles"), which can be serialized/deserialized into/from the "compiled model" as an IValue.
Also introduces APIs to register/look-up "compiled assembly code".

Differential Revision: [D27921866](https://our.internmc.facebook.com/intern/diff/D27921866/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27921866/)!

Differential Revision: [D27921866](https://our.internmc.facebook.com/intern/diff/D27921866)

[ghstack-poisoned]

ljk53 committed:
Update on "[pytorch][nnc] protocol classes to persist the context for compiled functions"




This is part of the changes to enable NNC AOT compilation for mobile.
At the end of the ahead-of-time compilation the compiler produces two sets of artifacts:
1. "compiled assembly code" - kernel functions in assembly format optimized for target platforms;
2. "compiled model" - regular TorchScript model that contains serialized parameters (weights/bias/etc) and invokes kernel functions via "handles" (name/version id/input & output specs/etc of the kernel functions).

This PR introduces a set of classes to represent kernel functions (a.k.a "handles"), which can be serialized/deserialized into/from the "compiled model" as an IValue.
Also introduces APIs to register/look-up "compiled assembly code".

Differential Revision: [D27921866](https://our.internmc.facebook.com/intern/diff/D27921866/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27921866/)!

Differential Revision: [D27921866](https://our.internmc.facebook.com/intern/diff/D27921866)

[ghstack-poisoned]
